feat(symbols): Add platform-restricted builtin symbol sources with org access control

vaind opened the pull request:
## Summary

Adds support for **platform-restricted builtin symbol sources** with organization-level access control.

This enables symbol sources to be restricted to specific platforms (e.g., console platforms like Nintendo Switch) and only accessible to organizations with the appropriate console platform access enabled.

**Companion PR**: [getsentry/getsentry#18867](https://github.com/getsentry/getsentry/pull/18867) - Defines the Nintendo symbol source configuration (credentials, bucket settings, etc.)

## Changes

### Platform Filtering for Builtin Sources API
- **[builtin_symbol_sources.py](src/sentry/api/endpoints/builtin_symbol_sources.py)**: Add platform-based filtering
  - Filter sources by `platform` query parameter
  - Check organization's `enabledConsolePlatforms` access for restricted sources
  - Backward compatible (works without platform parameter)

### Default Symbol Sources Enhancement
- **[default_symbol_sources.py](src/sentry/api/helpers/default_symbol_sources.py)**: Enhanced helper for platform-specific defaults
  - Added `nintendo-switch` platform mapping to `nintendo` source
  - Check platform restrictions and org access before adding sources
  - Organization auto-fetched from project if not provided

### Shared Utility
- **[console_platforms.py](src/sentry/utils/console_platforms.py)**: New utility module
  - `organization_has_console_platform_access()` - checks if org has access to a console platform
  - Used by both builtin sources endpoint and tempest utils

### Schema Update
- **[sources.py](src/sentry/lang/native/sources.py)**: Add `platforms` field to source schema

### Refactoring
- **[tempest/utils.py](src/sentry/tempest/utils.py)**: Refactored to use shared `organization_has_console_platform_access` utility
- **[team_projects.py](src/sentry/core/endpoints/team_projects.py)**: Pass organization to `set_default_symbol_sources()`

## How It Works

### Platform Restrictions

Sources can include a `platforms` key to restrict visibility:

```python
"nintendo": {
    "type": "s3",
    "platforms": ["nintendo-switch"],  # Only visible to nintendo-switch projects
    # ... other config (defined in getsentry)
}
```

**Two-layer filtering:**
1. **Platform match**: Request's platform must match source's `platforms` list
2. **Org access**: Organization must have platform in `enabledConsolePlatforms`

### Default Symbol Sources

```python
DEFAULT_SYMBOL_SOURCES = {
    "electron": ["ios", "microsoft", "electron"],
    "unity": ["ios", "microsoft", "android", "nuget", "unity", "nvidia", "ubuntu"],
    "unreal": ["ios", "microsoft", "android", "nvidia", "ubuntu"],
    "godot": ["ios", "microsoft", "android", "nuget", "nvidia", "ubuntu"],
    "nintendo-switch": ["nintendo"],
}
```

## Test Plan

- [x] Unit tests pass
- [x] Pre-commit hooks pass
- [x] Mypy type checking passes
- [x] Verify Nintendo Switch project in enabled org sees the source
- [x] Verify Nintendo Switch project in non-enabled org does NOT see the source
- [x] Verify non-Nintendo projects never see the source
- [x] Verify public sources still visible to all platforms
- [x] Verify API works without platform parameter (backward compatibility)

vaind committed:
feat(symsorter): Add platform-restricted builtin symbol sources with org access control

Adds support for restricting builtin symbol sources to specific platforms with
organization-level access control via enabledConsolePlatforms.

## Changes

- Add `platforms` key to builtin source definitions in server.py to restrict
  sources to specific platforms (e.g., nintendo-switch)
- Update builtin_symbol_sources endpoint to filter sources based on:
  1. Platform match: request platform must match source's platforms list
  2. Org access: organization must have console platform enabled
- Pass project platform from frontend to API for filtering
- Add nintendo-private S3 source with credentials from environment variables
- Add platform mapping for nintendo-switch in DEFAULT_SYMBOL_SOURCES

## Implementation Details

Sources without a `platforms` key remain visible to all platforms (backward compatible).
Sources with `platforms` are only visible when:
- The request includes a matching platform parameter AND
- The organization has that console platform in enabledConsolePlatforms

This ensures platform-restricted sources (like Nintendo's private symbol server)
are only accessible to authorized organizations with matching project platforms.

vaind committed:
chore: reorder fields

vaind committed:
formatting

vaind committed:
Update static/app/views/settings/projectDebugFiles/index.tsx

Co-authored-by: Priscila Oliveira <priscilawebdev@gmail.com>

vaind committed:
feat(symsorter): Add platform-restricted builtin symbol sources with org access control

Adds support for platform-restricted symbol sources (e.g., nintendo-private) with
organization-level access control for console platforms.

Changes:
- Add 'platforms' field to COMMON_SOURCE_PROPERTIES schema for validation
- Add "nintendo-switch": ["nintendo-private"] to DEFAULT_SYMBOL_SOURCES mapping
- Update set_default_symbol_sources() to check organization access for platform-restricted sources
  - Checks source's 'platforms' field against org's 'enabled_console_platforms' option
  - Only adds source if org has access to at least one required platform
  - Supports both Django Project and RpcProject types
- Pass organization parameter in apply_default_project_settings()
- Add comprehensive test suite with 8 test cases

How it works:
- For nintendo-switch projects, checks if org has "nintendo-switch" in enabled_console_platforms
- If yes: adds nintendo-private source
- If no: skips nintendo-private source
- Non-restricted sources (ios, microsoft, etc.) are added without access checks

Benefits:
- Explicit platform mapping in DEFAULT_SYMBOL_SOURCES (follows existing pattern)
- Organization access control for console platform sources
- Scalable to other console platforms (PlayStation, Xbox, etc.)
- Automatic org fetch from project if not provided

Test coverage:
- Basic platform handling (no platform, unknown, electron, unity)
- Organization auto-fetch from project
- Platform-restricted sources with and without org access
- Non-restricted sources unaffected by console platform access

vaind committed:
Merge branch 'master' into feat/platform-restricted-symbol-sources

vaind committed:
cleanup

vaind committed:
Merge remote-tracking branch 'upstream/master' into feat/platform-restricted-symbol-sources

vaind committed:
fix: Preserve non-existent symbol sources in platform defaults

The original implementation incorrectly filtered out symbol sources like
"ios" and "android" that don't exist in SENTRY_BUILTIN_SOURCES. These
sources are intentionally included in defaults even though they don't
exist in the config - they get filtered at runtime in sources.py.

Changes:
- Only check platform restrictions for sources that exist in config
- Sources without config entries are included without restriction checks
- Always update project option for recognized platforms (even if empty)
- Fix tests to verify epoch defaults are preserved for unknown platforms

This ensures the existing test_team_projects.py tests continue to pass
while properly implementing platform-restricted sources.

vaind committed:
Merge remote-tracking branch 'upstream/master' into feat/platform-restricted-symbol-sources

vaind committed:
fix(types): Add type casts for platform restrictions to fix mypy errors

Use typing.cast to explicitly cast source.get('platforms') to list[str] | None
to satisfy mypy's type checking. The cast is safe because we check if the
source exists before accessing it, and the platforms field is always either
None or list[str] per the SENTRY_BUILTIN_SOURCES configuration.

vaind committed:
chore: Remove frontend changes (will be in separate PR)

vaind committed:
perf: Address Seer bot review feedback

- Move organization.get_option() call outside loop to avoid repeated DB calls
- Add error handling for Organization.DoesNotExist when fetching from RpcProject
- Add comment explaining optimization

Addresses HIGH and MEDIUM severity issues identified by Seer bot review.

loewenheim commented on src/sentry/api/helpers/default_symbol_sources.py:
```diff
@@ -7,11 +13,69 @@
+
+        # Include the source (either it passed platform check or doesn't exist in config)
+        # Non-existent sources will be filtered out at runtime in sources.py
+        enabled_sources.append(source_key)
```
Why are we appending nonexistent sources here?

loewenheim commented on src/sentry/conf/server.py:
```diff
@@ -2438,6 +2438,19 @@ def custom_parameter_sort(parameter: dict) -> tuple[str, int]:
+        "region": os.environ.get("NINTENDO_SYMBOL_SERVER_S3_REGION", ""),
+        "access_key": os.environ.get("NINTENDO_SYMBOL_SERVER_AWS_ACCESS_KEY_ID", ""),
+        "secret_key": os.environ.get("NINTENDO_SYMBOL_SERVER_AWS_SECRET_ACCESS_KEY", ""),
+        "prefix": os.environ.get("NINTENDO_SYMBOL_SERVER_S3_PREFIX", ""),
```
`getsentry` PR for the symbol source: https://github.com/getsentry/getsentry/pull/18867

loewenheim commented on src/sentry/conf/server.py:
```diff
@@ -2438,6 +2438,19 @@ def custom_parameter_sort(parameter: dict) -> tuple[str, int]:
+        "access_key": os.environ.get("NINTENDO_SYMBOL_SERVER_AWS_ACCESS_KEY_ID", ""),
+        "secret_key": os.environ.get("NINTENDO_SYMBOL_SERVER_AWS_SECRET_ACCESS_KEY", ""),
+        "prefix": os.environ.get("NINTENDO_SYMBOL_SERVER_S3_PREFIX", ""),
+    },
```
```suggestion
```

loewenheim reviewed (approved):
If you remove the definition of the Nintendo source (and change all references to `nintendo-private` to `nintendo`, in accordance with the `getsentry` PR, this is ready to merge.

vaind committed:
Apply suggestions from code review

Co-authored-by: Sebastian Zivota <loewenheim@users.noreply.github.com>

vaind commented on tests/sentry/api/endpoints/test_builtin_symbol_sources.py:
```diff
@@ -1,5 +1,33 @@
+    },
+    "nintendo": {
+        "id": "sentry:nintendo",
+        "name": "Nintendo Private",
```
```suggestion
        "name": "Nintendo SDK",
```

vaind committed:
Apply suggestions from code review

vaind committed:
Merge branch 'master' into feat/platform-restricted-symbol-sources

priscilawebdev commented on src/sentry/api/helpers/default_symbol_sources.py:
```diff
@@ -7,11 +13,69 @@
     "unity": ["ios", "microsoft", "android", "nuget", "unity", "nvidia", "ubuntu"],
     "unreal": ["ios", "microsoft", "android", "nvidia", "ubuntu"],
     "godot": ["ios", "microsoft", "android", "nuget", "nvidia", "ubuntu"],
+    "nintendo-switch": ["nintendo"],
```
this is a valid feedback. shouldn't we name it "nintendo-private" to match the key in `SENTRY_BUILTIN_SOURCES`?

priscilawebdev commented on tests/sentry/api/endpoints/test_builtin_symbol_sources.py:
```diff
@@ -1,5 +1,33 @@
+        "platforms": ["nintendo-switch"],
+    },
+}
+
```
instead of defining this object, can't we just import the `SENTRY_BUILTIN_SOURCES`?

priscilawebdev commented on tests/sentry/api/endpoints/test_builtin_symbol_sources.py:
```diff
@@ -39,3 +67,77 @@ def test_with_slug(self) -> None:
+        assert "nintendo" not in source_keys
+        # Should still see public sources
+        assert "public-source-1" in source_keys
+        assert "public-source-2" in source_keys
```
are we adding an additional security check here? Users should only be able to create a Nintendo Switch project if they have access. Or is the idea that access could later be revoked, in which case they would already have an existing Nintendo Switch project?

priscilawebdev commented on tests/sentry/api/helpers/test_default_symbol_sources.py:
```diff
@@ -0,0 +1,110 @@
+
+        # Should not change the defaults for projects with unknown platforms
+        sources = project.get_option("sentry:builtin_symbol_sources")
+        assert sources == epoch_defaults
```
should we test if the epoch changes, if a known platform is provided?

vaind replied:
This is already covered by the existing tests:
test_electron_platform (line 51) - verifies electron gets ios, microsoft, electron
test_unity_platform (line 62) - verifies unity gets ios, microsoft, android, nuget, unity, nvidia, ubuntu
These tests implicitly verify epoch defaults are overridden because they assert platform-specific sources are present.

vaind committed:
feat(symbol-sources): remove Nintendo Symbol Server configuration

vaind committed:
refactor(symbol-sources): extract console platform access check to shared utility

Move organization_has_console_platform_access to sentry/utils/console_platforms.py
so it can be reused by both builtin_symbol_sources.py and tempest/utils.py.
Also fix platform-restricted tests to use mock settings since nintendo source
is defined in getsentry, not in the default SENTRY_BUILTIN_SOURCES.

vaind committed:
Merge branch 'master' into feat/platform-restricted-symbol-sources

priscilawebdev reviewed (approved):
LGTM! 👏

mujacica reviewed (approved):
LGTM
